### ISSUE TITLE
wesl.toml for LSPs

### ISSUE THREAD

**Issue by stefnotch** [wgsl-tooling-wg/wesl-spec#57]
We need a way for language servers to figure out what file an "absolute path" resolves to.
As in, if I write `import my_project::foo`, which folder does the language server look at?

## `wesl.toml` Proposal

First, a language server scans for `wesl.toml` files (how do they do it for Cargo.toml and package.json? Do they only scan the root folder, or..?)

The file then contains a list of dependencies. For each dependency, we have its name, and how to find it. Currently, the only valid option is with a path that is relative to the `wesl.toml`.

e.g. 
```toml
[dependencies] # Or suggest a different name
my_lib = { path = "./src/lib.wesl" } 
```

The importing proposal #53 then uses these entries to start fully qualified paths. 

### Example
```
import my_lib::foo::bar;
```
1. `my_lib` is found in the `wesl.toml`. Valid start
2. The fully qualified path then got resolved to `["my_lib", "bar", "foo.wesl"]`. 

### Implementation detail
*Internally*, a bundler should also generate an absolute filepath that is relative to the `wesl.toml`. Why?
If the user specifies
```
my_lib = { path = ./src/lib.wesl } 
utils = { path = ./src/shaders/utils.wesl }
```
Then utils my_lib/shaders/utils and utils should have the same absolute path. That way, the linker only needs to look at the absolute path to find out whether it's the same module or not.

**Comment by ncthbrt** [comment]
This makes a lot of sense. Though we still need to figure out how this might work with bevy's cargo shaders

**Comment by k2d222** [comment]
agreed with the proposal and @ncthbrt comment. I think it's time to try some prototypes to depend on npm/cargo packages.
I would also like if the wesl.toml file was optional. An implementation could fallback to looking at the cargo.toml/package.json.

**Comment by stefnotch** [comment]
@k2d222 This current wesl.toml proposal is purely for having non-relative imports within a project. Akin to `import { cat } from "@/foo"` or `use crate::foo::cat`.

As long as one only uses relative imports, this wesl.toml wouldn't be needed.

I'm somewhat sceptical about making it the libraries part optional, because from the point of view of a language server, it's not entirely obvious which files and package managers it should check. Bonus points for complex cases like monorepos with multiple package.json files.
But I also get that adding a config in two separate files is annoying. Maybe a good-enough compromise could be
1. wesl.tom file is not optional
2. wesl.toml file can "extend" another file. If I extend my package.json, then every library from there can be used in wesl.

**Comment by stefnotch** [comment]
And yes, we should definitely experiment with npm/cargo packages! :)

**Comment by k2d222** [comment]
> I'm somewhat sceptical about making it the libraries part optional, because from the point of view of a language server, it's not entirely obvious which files and package managers it should check. 

99% of cases are not edge-cases, let's optimize UX for that. in those cases the `package.json`/`cargo.toml` file already contains all the information and the wesl file would be redundant. I don't think it would be too difficult for the langserver either. With your compromise, it would already have to know how to parse package.json/cargo.toml files.

**Comment by mighdoll** [comment]
(See also #29 for some related discussions)

**Comment by mighdoll** [comment]
See also #30 to discuss library format more generally.

**Comment by mighdoll** [comment]
Copied from #29: 
> Summarizing because it's been a while and ideas may have drifted:
> 
>     * we like the TOML syntax better
> 
>     * we want to standardize that TOML file accross Rust and JavaScript
> 
>     * build-time linkers (either via the vite plugin, `build.rs`  or a CLI) are responsible for parsing the TOML file and generate appropriate code to invoke the linker.
> 

Agreed. 

I think a compatibility goal would be: users should be able to seamlessly copy a shader files from a JavaScript project to a Rust project w/o changing the wesl or toml files. 

>     * projects not using Vite can write the JS/Rust code by hand, should be easy enough.

Yep things must work w/o a bundler. 

Also noting for future readers that wesl-plugin is available for many other bundlers too: webpack, nuxt, rollup, farm, rspack, esbuild, etc.
 
> My assumptions:

>     * published packages _Must_ contain the TOML file (maybe bundlers could _optionally_ auto-generate that file if a `package.json` or `Cargo.toml` file containing necessary metadata is nearby).
> 
>     * application code (not packages) _Can_ contain the TOML file, but can also choose to configure the linker manually. 

I see that wesl libraries should have a conformant toml file in their source on github, so that a packaging tool can create npm/cargo packages.

I'm fuzzy on where else the toml file needs to be. Is the toml file is needed for the language server for app code? Does the language server need the toml file for libraries? Are npm/cargo packages readable cross ecosystem?

**Comment by mighdoll** [comment]
I think Deno has bowed to this reality too, no? 

I guess it's hard to change the world, even if you invented it.

**Comment by stefnotch** [comment]
> I think Deno has bowed to this reality too, no?
> 
> I guess it's hard to change the world, even if you invented it.

IIRC, it only creates a node_modules folder for packages from npm.

**Comment by mighdoll** [comment]
> listing directly referenced libraries by name. e.g. weslLibs = ["random_wesl"]. Not sure if that's needed yet.

I was able to find directly referenced libraries by scanning the sources for import statements. So explicit library references aren't needed so far as I've found.

**Comment by mighdoll** [comment]
mention in spec that there will be a wesl.toml format defined later, and point to wesl-js docs noting that wesl.toml is required there

**Comment by mighdoll** [comment]
The current format in wesl-js looks like this:

```toml
weslFiles = [ "./shaders/**/*.wesl" ]
weslRoot = "./shaders"
dependencies = ["random_wgsl"]
```

The schema is defined [here](https://github.com/wgsl-tooling-wg/wesl-js/blob/master/tools/packages/wesl-plugin/src/WeslPlugin.ts#L24):

```ts
export interface WeslToml {
  /** glob search strings to find .wesl/.wgsl files. Relative to the toml directory. */
  weslFiles: string[];

  /** base directory for wesl files. Relative to the toml directory. */
  weslRoot: string;

  /** names of directly referenced wesl shader packages (e.g. npm package names) */
  dependencies?: string[];
}
```

The wesl.toml is used by developer tools like the library packaging tool and bundler plugins. wesl.toml is not intended to be used at runtime in wesl-js. 

To find the code/metadata for wesl package dependencies, a tool can 'resolve' the npm package name using node package resolution. There are libraries that do reliable node package resolution available in js and rust.

**Comment by mighdoll** [comment]
Added some docs to wesl-lang.dev. 
https://wesl-lang.dev/docs/Getting-Started-JavaScript#wesltoml

**Comment by mighdoll** [comment]
moving to M2 for future work on wesl.toml format
